### PR TITLE
[ERE-882] Build MIMS integration capability

### PR DESCRIPTION
**Pull request by alintreznai**
Related TRRF PR: https://github.com/eresearchqut/trrf/pull/288

### CODE REVIEW

File: mnd/mnd/integration/mims/mims_view.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+)
+
+logger = logging.getLogger(__name__)
+
+
+@require_GET
+def product_search(request):
+
+    def active_ingredient(result):
+        return ", ".join(result.split("+"))
```
Comment (ScriptSmith) [inline mnd/mnd/integration/mims/mims_view.py]: Could be split by " + " to avoid gaps before commas

File: mnd/mnd/integration/mims/mims_view.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        if document_link and document_link[0]:
+            for p in products:
+                update_cmi_cache(CMIInfo(p['productId'], p['productName'], cmi_id, document_link[0], True))
+            cmi_info = update_cmi_cache(CMIInfo(product_id, '', cmi_id, document_link[0], True))
+            return cmi_info
+        return update_cmi_cache(CMIInfo(product_id, product_name, cmi_id, '', False))
+
+
+@require_GET
+def cmi_details(request):
```
Comment (ScriptSmith) [inline mnd/mnd/integration/mims/mims_view.py:32]: I will check with Yvette when she gets back, but we may need to proxy + cache the cmi document as well

File: mnd/mnd/forms/widgets/mnd_widgets.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+from django.urls import reverse_lazy
```
Comment (ScriptSmith) [inline mnd/mnd/forms/widgets/mnd_widgets.py:1]: Similar bug as with multi-cdefiles where if you navigate away from a saved form it asks for confirmation

File: mnd/mnd/integration/mims/mims_view.py (hunk 1 above)
Comment (alintreznai) [reply mnd/mnd/integration/mims/mims_view.py]: Updated.

File: mnd/mnd/integration/mims/mims_api.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                "api-key": self.api_key,
+                "Authorization": f"Bearer {token}"
+            }
+        except KeyError:
+            self._refresh_token()
+
+    def _full_url(self, endpoint):
+        return f"{self.service_endpoint}/{endpoint}"
+
+    def search_brand(self, brand, page=1, limit=50):
```
Comment (ScriptSmith) [inline mnd/mnd/integration/mims/mims_api.py]: Not used. Intended for use in future widgets?

File: mnd/mnd/integration/mims/mims_api.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        params = urllib.parse.urlencode({
+            "term": brand,
+            "include": True,
+            "page": page,
+            "limit": limit
+        })
+        resp = requests.get(self._full_url(f"{self.BRAND_URI}?{params}"), headers=self._make_auth_header())
+        return resp.json()
+
+    def get_brand_details(self, brand_id):
```
Comment (ScriptSmith) [inline mnd/mnd/integration/mims/mims_api.py]: Not used

File: mnd/mnd/static/css/form_common.css [hunk 6]
```diff
@@ -0,0 +1,27 @@
+.ui-autocomplete {
```
Comment (ScriptSmith) [inline mnd/mnd/static/css/form_common.css]: Are these changes just for the autocomplete? They look too broad if that's the case

File: mnd/mnd/forms/widgets/mnd_widgets.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    def render(self, name, value, attrs, renderer=None):
+        return """
+            <input type="text" name="%s_" id="id_%s"/>
+            <input type="hidden" name="%s" id="id_%s_" value="%s"/>
+            <div>
+                <label>Active ingredients</label><textarea id="id_%s_ingredients" disabled></textarea>
+                <label>Therapeutic classes</label><textarea id="id_%s_mims" disabled> </textarea>
+            </div>
+            <script type="text/javascript">
+                lookupValue($("#id_%s"), '%s', '%s', ['ingredients', 'mims']);
```
Comment (ScriptSmith) [inline mnd/mnd/forms/widgets/mnd_widgets.py:24]: This causes a lookup even when no value has been set for the field, which in turn means that the exclamation mark appears when the form is first loaded for a patient

File: mnd/mnd/integration/mims/mims_service.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    def active_ingredient(result):
+        return ", ".join(result.split(" + "))
+
+    if product and product.strip() != '' and len(product) >= 4:
+        cached = search_cache(product)
+        if cached:
+            return [ProductSearchResult(r.id, r.name, r.activeIngredient) for r in cached]
+
+        result = api.search_product(product)
```
Comment (ScriptSmith) [inline mnd/mnd/integration/mims/mims_service.py]: I see there are pagination fields in the API. Do we need to consider the edge case where a common drug has so many variants that turn up in a search, that we need to paginate over it to provide a complete list to the user?

Or maybe there needs to be a visual indication that the results aren't complete if there are more pages?

File: mnd/mnd/integration/mims/mims_api.py (hunk 4 above)
Comment (alintreznai) [reply mnd/mnd/integration/mims/mims_api.py]: Removed.

File: mnd/mnd/static/css/form_common.css (hunk 6 above)
Comment (alintreznai) [reply mnd/mnd/static/css/form_common.css]: Removed. Set the change only in the form view.

File: mnd/mnd/forms/widgets/mnd_widgets.py (hunk 7 above)
Comment (alintreznai) [reply mnd/mnd/forms/widgets/mnd_widgets.py:24]: Updated implementation of lookupValue to not fetch data if there is no value

File: mnd/mnd/integration/mims/mims_service.py (hunk 8 above)
Comment (alintreznai) [reply mnd/mnd/integration/mims/mims_service.py]: Added pagination support when searching products.

File: mnd/mnd/integration/mims/mims_api.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    def search_product(self, product, page, limit=PAGE_SIZE):
+        params = urllib.parse.urlencode({
+            "term": product,
+            "include": True,
+            "page": page,
+            "limit": limit
+        })
+        resp = requests.get(self._full_url(f"{self.PRODUCT_URI}?{params}"), headers=self._make_auth_header())
+        if resp.status_code == 204:
```
Comment (ScriptSmith) [inline mnd/mnd/integration/mims/mims_api.py]: Redundant because of line 66

Comment (alintreznai) [reply mnd/mnd/integration/mims/mims_api.py]: Removed.

File: mnd/mnd/integration/mims/mims_api.py (hunk 5 above)
Comment (alintreznai) [reply mnd/mnd/integration/mims/mims_api.py]: Removed.

**Pull request merged**